### PR TITLE
Shed Tail AI + Small Substitute AI fix

### PR DESCRIPTION
**Pull request by SarnPoke**
## Description

Adds Shed Tail AI checks to `AI_CheckViability` and `AI_CheckBadMove`. Also fixes perceived syntax error in Substitute AI.

## Issue(s) that this PR fixes

#5238

## People who collaborated with me in this PR
@AlexOn1ine, @Pawkkie 

## **Discord contact info**

sarnnn

### CODE REVIEW

**Review by AlexOn1ine — COMMENTED** [review]
Since both Sub and Shed Tail checks are very similar I would make a function in `battle_ai_util.c` and handle the differences inside it to remove duplicate code.

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -3591,7 +3601,33 @@ static u32 AI_CalcMoveEffectScore(u32 battlerAtk, u32 battlerDef, u32 move)
...
           || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
             ADJUST_SCORE(GOOD_EFFECT);
-        if (!gBattleMons[battlerDef].status2 & (STATUS2_WRAPPED | STATUS2_ESCAPE_PREVENTION && aiData->hpPercents[battlerAtk] > 70))
+        if (!(gBattleMons[battlerDef].status2 & (STATUS2_WRAPPED | STATUS2_ESCAPE_PREVENTION)) && aiData->hpPercents[battlerAtk] > 70) // Why are we doing this?
+            ADJUST_SCORE(WEAK_EFFECT);
+        break;
+    case EFFECT_SHED_TAIL: // Mostly the same as the AI for Substitute
+        if (ShouldSwitch(battlerAtk, FALSE) 
+          && (gStatuses3[battlerAtk] & (STATUS3_ROOTED | STATUS3_AQUA_RING | STATUS3_MAGNET_RISE | STATUS3_POWER_TRICK) || AnyStatIsRaised(battlerAtk))
+          && (aiData->hpPercents[battlerAtk] > 50) // Just extra reinforcement that this should never be used when at 50 or less HP
```
[the remark this answers is not in the document]
Comment (SarnPoke) [reply src/battle_ai_main.c]: Agreed, redundant checks have been removed.

File: src/battle_ai_main.c [hunk 2]
```diff
@@ -3591,7 +3601,33 @@ static u32 AI_CalcMoveEffectScore(u32 battlerAtk, u32 battlerDef, u32 move)
           || HasMoveEffect(battlerDef, EFFECT_CONFUSE)
           || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
             ADJUST_SCORE(GOOD_EFFECT);
-        if (!gBattleMons[battlerDef].status2 & (STATUS2_WRAPPED | STATUS2_ESCAPE_PREVENTION && aiData->hpPercents[battlerAtk] > 70))
+        if (!(gBattleMons[battlerDef].status2 & (STATUS2_WRAPPED | STATUS2_ESCAPE_PREVENTION)) && aiData->hpPercents[battlerAtk] > 70) // Why are we doing this?
```
[the remark this answers is not in the document]
Comment (SarnPoke) [reply src/battle_ai_main.c]: Removed.

**Comment by SarnPoke** [comment]
> Since both Sub and Shed Tail checks are very similar I would make a function in `battle_ai_util.c` and handle the differences inside it to remove duplicate code.

This makes sense, yeah. I've done this now.

**Comment by SarnPoke** [comment]
Not sure why this test is failing now, when it didn't before:
```c++
AI_SINGLE_BATTLE_TEST("AI prefers to keep its substitute over removing hazards if target is slower")
{
    GIVEN {
        AI_FLAGS(AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_CHECK_VIABILITY | AI_FLAG_TRY_TO_FAINT);
        PLAYER(SPECIES_WOBBUFFET) { Speed(50); Status1(STATUS1_PARALYSIS); Moves(MOVE_SLEEP_POWDER, MOVE_STEALTH_ROCK, MOVE_CELEBRATE); }
        OPPONENT(SPECIES_WOBBUFFET) { Speed(100); Moves(MOVE_BITE, MOVE_TACKLE, MOVE_SUBSTITUTE, MOVE_TIDY_UP); }
    } WHEN {
        TURN { MOVE(player, MOVE_STEALTH_ROCK); EXPECT_MOVE(opponent, MOVE_TIDY_UP); }
        TURN { MOVE(player, MOVE_STEALTH_ROCK); EXPECT_MOVE(opponent, MOVE_SUBSTITUTE); }
        TURN { EXPECT_MOVE(opponent, MOVE_BITE); }
    }
}
```

File: src/battle_ai_util.c [hunk 3]
```diff
@@ -3987,3 +3987,40 @@ bool32 AI_ShouldSpicyExtract(u32 battlerAtk, u32 battlerAtkPartner, u32 move, st
...
+
+    if (HasMoveEffect(battlerDef, EFFECT_SLEEP)
+    || HasMoveEffect(battlerDef, EFFECT_TOXIC)
+    || HasMoveEffect(battlerDef, EFFECT_POISON)
+    || HasMoveEffect(battlerDef, EFFECT_PARALYZE)
+    || HasMoveEffect(battlerDef, EFFECT_WILL_O_WISP)
+    || HasMoveEffect(battlerDef, EFFECT_CONFUSE)
+    || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
+        ADJUST_SCORE_PTR(GOOD_EFFECT);
+}
```
Comment (Pawkkie) [inline src/battle_ai_util.c:4026]:
```suggestion
    if (aiData->hpPercents[battlerAtk] > 70)
        ADJUST_SCORE_PTR(WEAK_EFFECT);
}
```

**Review by Pawkkie — CHANGES_REQUESTED** [review]
The test was failing because it relies on this check:
```
if (aiData->hpPercents[battlerAtk] > 70)
    ADJUST_SCORE_PTR(WEAK_EFFECT);
```
which makes Substitute ever so slightly preferable to other status moves when the mon is at higher HP thresholds. I completely agree with axing the status2 half, but this half is what makes the Tidy Up behaviour work properly :)

I tried to make a proper "review suggestion" with this change in it which I've never done before let's see if it worked lol

File: src/battle_ai_main.c [hunk 4]
```diff
@@ -3572,27 +3582,10 @@ static u32 AI_CalcMoveEffectScore(u32 battlerAtk, u32 battlerDef, u32 move)
...
-          || HasMoveEffect(battlerDef, EFFECT_WILL_O_WISP)
-          || HasMoveEffect(battlerDef, EFFECT_CONFUSE)
-          || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
-            ADJUST_SCORE(GOOD_EFFECT);
-        if (!gBattleMons[battlerDef].status2 & (STATUS2_WRAPPED | STATUS2_ESCAPE_PREVENTION && aiData->hpPercents[battlerAtk] > 70))
-            ADJUST_SCORE(WEAK_EFFECT);
+        IncreaseSubstituteMoveScore(battlerAtk, battlerDef, move, &score);
+        break;
+    case EFFECT_SHED_TAIL:
+        IncreaseSubstituteMoveScore(battlerAtk, battlerDef, move, &score);
```
Comment (AlexOn1ine) [inline src/battle_ai_main.c:3588]:
```suggestion
    case EFFECT_SUBSTITUTE:
    case EFFECT_SHED_TAIL:
        IncreaseSubstituteMoveScore(battlerAtk, battlerDef, move, &score);
```

File: src/battle_ai_main.c [hunk 5]
```diff
@@ -3572,27 +3582,10 @@ static u32 AI_CalcMoveEffectScore(u32 battlerAtk, u32 battlerDef, u32 move)
         IncreaseParalyzeScore(battlerAtk, battlerDef, move, &score);
         break;
     case EFFECT_SUBSTITUTE:
```
Comment (AlexOn1ine) [inline src/battle_ai_main.c:3584]:
```suggestion
```

**Review by Pawkkie — APPROVED** [review]

File: src/battle_ai_util.c [hunk 6]
```diff
@@ -3987,3 +3987,43 @@ bool32 AI_ShouldSpicyExtract(u32 battlerAtk, u32 battlerAtkPartner, u32 move, st
...
+    // if (IsPredictedToSwitch(battlerDef, battlerAtk)
+    //     ADJUST_SCORE_PTR(DECENT_EFFECT);
+
+    if (HasMoveEffect(battlerDef, EFFECT_SLEEP)
+    || HasMoveEffect(battlerDef, EFFECT_TOXIC)
+    || HasMoveEffect(battlerDef, EFFECT_POISON)
+    || HasMoveEffect(battlerDef, EFFECT_PARALYZE)
+    || HasMoveEffect(battlerDef, EFFECT_WILL_O_WISP)
+    || HasMoveEffect(battlerDef, EFFECT_CONFUSE)
+    || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:4024]:
```suggestion
     || HasMoveEffect(battlerDef, EFFECT_TOXIC)
     || HasMoveEffect(battlerDef, EFFECT_POISON)
     || HasMoveEffect(battlerDef, EFFECT_PARALYZE)
     || HasMoveEffect(battlerDef, EFFECT_WILL_O_WISP)
     || HasMoveEffect(battlerDef, EFFECT_CONFUSE)
     || HasMoveEffect(battlerDef, EFFECT_LEECH_SEED))
```

**Pull request closed without merging**